metadata-models/src/main/pegasus/com/linkedin/chart/ChartQuery.pdl

```diff
@@ -0,0 +1,33 @@
+namespace com.linkedin.chart
+
+/**
+ * Information for chart query which is used for getting data of the chart
+ */
+record ChartQuery {
+
+  /**
+   * Raw query to build a chart from input datasets
+   */
+  rawQuery: string
+
+  /**
+   * Chart query type
+   */
+  type: enum ChartQueryType {
+
+    /**
+     * LookML queries
+     */
+    LOOKML
```

Comment:
I'm sort of concerned about enums like this when it comes to open source. They're a little hard for other companies to expand, and could get quite quickly overloaded. I know we have a bunch already, but I figured now may be a good time to bring it up?

It might be better to instead treat ChartQuery as abstract, and have specific `SqlQuery` and `LookMlQuery` records that extend `ChartQuery`? Let the actual type determine the type?

Reply:
Thanks for the suggestion. But, your suggestion is a little tricky to implement in PDSC/PDL world where there is no true inheritance support like you described. There is support for `includes` which we tried to treat as `inheritance` in a couple of places but that actually is more `composition` than `inheritance` :)

And there is no true polymorphism as well meaning: I can't still keep the type of `query` field as `ChartQuery` in `ChartInfo` model and expect to set/get `SqlQuery` and `LookMlQuery` to/out of it.

Reply:
Ah, yeah, I guess that's why we use unions, isn't it :/

So just to float the idea, not that I like it, is to make this a string instead. While this is more extensible, it is harder for people to enforce a common set of strings (`looker` vs `Looker` vs `LOOKER`). Worth noting that types in the proposed URNs are strings.

Here's a thought experiment: if we (LinkedIn) wanted to use these open source models with some internal infrastructure, how would we easily modify and/or extend them? Keeping a fork of this file internally isn't that easy; you have to have something in place to prevent having this go upstream again.

Reply:
I understand that this would make things more extensible if we just make this a simple string. But, great extensibility/flexibility comes with the cost of great responsibility :) In this case, this will not be strongly-typed and it's impossible to assure what is being put to this field unless clients do that by themselves.

Not that it's the best approach but you can approach modelling for internal use as below (if model forking is not an option):

Define another model to accommodate internal only query types as:

```
enum LIOnlyChartQuery {
  proprieratyQuery1
  proprieratyQuery2
}
```

```
typeref LIChartQuery = union[ChartQuery, LiOnlyChartQuery]
```

WDYT?